src/mapper/pkg/resolvers/intents_store.go

```diff
@@ -0,0 +1,130 @@
+package resolvers
+
+import (
+	"encoding/json"
+	"github.com/otterize/network-mapper/src/mapper/pkg/graph/model"
+	"github.com/samber/lo"
+	"sort"
+	"time"
+)
+
+type ServiceToDestToTimestamp map[model.OtterizeServiceIdentity]map[model.OtterizeServiceIdentity]time.Time
+
+type DiscoveredIntent struct {
+	Source      model.OtterizeServiceIdentity
+	Destination model.OtterizeServiceIdentity
+	Timestamp   time.Time
+}
+
+type IntentsHolderStore struct {
+	serviceMap ServiceToDestToTimestamp
+}
+
+func NewIntentsHolderStore() IntentsHolderStore {
+	return IntentsHolderStore{
+		serviceMap: make(ServiceToDestToTimestamp),
+	}
+}
+
+func (serviceMap ServiceToDestToTimestamp) MarshalJSON() ([]byte, error) {
```

Comment:
There's multiple levels of JSON marshalling going on - let's try to keep just one of these. You can implement MarshalJSON for the store and put this implementation there. Having all these different parts is confusing to read - there are too many levels of indirection, you need to keep mental state of intentsHolder -> intentsHolderStore -> this function to understand what happens when the store is serialized.

Try to reduce this to just 1 or 2 - implement the logic for MarshalJSON and UnmarshalJSON in the store itself, so they directly modify the map.

When should you implement Marshal/Unmarshal on a type like a map? When you intend to also reuse it elsewhere, and need the same implementation.  This is not the case here - it's only marshalled once.